utils/reconnecting_udp_conn_test.go

```diff
@@ -0,0 +1,477 @@
+	hostPort := "blahblah:34322"
+
+	mockServer, clientConn, err := newUDPConn()
+	require.NoError(t, err)
+	defer mockServer.Close()
+
+	resolver := mockResolver{}
+	resolver.
+		On("ResolveUDPAddr", "udp", hostPort).
+		Return(nil, fmt.Errorf("failed to resolve")).Once().
+		On("ResolveUDPAddr", "udp", hostPort).
+		Return(&net.UDPAddr{
+			IP:   net.IPv4(123, 123, 123, 123),
+			Port: 34322,
+		}, nil)
+
+	dialer := mockDialer{}
+	dialCall := dialer.
+		On("DialUDP", "udp", (*net.UDPAddr)(nil), &net.UDPAddr{
+			IP:   net.IPv4(123, 123, 123, 123),
+			Port: 34322,
+		}).
+		Return(clientConn, nil).Once()
+
+	conn, err := newReconnectingUDPConn(hostPort, time.Millisecond*10, resolver.ResolveUDPAddr, dialer.DialUDP, log.NullLogger)
+	assert.NoError(t, err)
+	require.NotNil(t, conn)
+
+	err = conn.SetWriteBuffer(UDPPacketMaxLength)
+	assert.NoError(t, err)
+
+	wasCalled := waitForCallWithTimeout(dialCall)
+	assert.True(t, wasCalled)
+
+	connEstablished := waitForConnCondition(conn, func(conn *reconnectingUDPConn) bool {
+		return conn.conn != nil
+	})
+
+	assert.True(t, connEstablished)
+
+	assertConnWritable(t, conn, mockServer)
+	assertSockBufferSize(t, UDPPacketMaxLength, clientConn)
+
+	err = conn.Close()
+	assert.NoError(t, err)
+
+	// assert the actual connection was closed
+	assert.Error(t, clientConn.Close())
+
+	resolver.AssertExpectations(t)
+	dialer.AssertExpectations(t)
+}
+
+func TestResolvedUDPConnNoSwapIfFail(t *testing.T) {
+	hostPort := "blahblah:34322"
+
+	mockServer, clientConn, err := newUDPConn()
+	require.NoError(t, err)
+	defer mockServer.Close()
+
+	resolver := mockResolver{}
+	resolver.
+		On("ResolveUDPAddr", "udp", hostPort).
+		Return(&net.UDPAddr{
+			IP:   net.IPv4(123, 123, 123, 123),
+			Port: 34322,
+		}, nil).Once()
+
+	failCall := resolver.On("ResolveUDPAddr", "udp", hostPort).
+		Return(nil, fmt.Errorf("resolve failed"))
+
+	dialer := mockDialer{}
+	dialer.
+		On("DialUDP", "udp", (*net.UDPAddr)(nil), &net.UDPAddr{
+			IP:   net.IPv4(123, 123, 123, 123),
+			Port: 34322,
+		}).
+		Return(clientConn, nil).Once()
+
+	conn, err := newReconnectingUDPConn(hostPort, time.Millisecond*10, resolver.ResolveUDPAddr, dialer.DialUDP, log.NullLogger)
+	assert.NoError(t, err)
+	require.NotNil(t, conn)
+
+	wasCalled := waitForCallWithTimeout(failCall)
+
+	assert.True(t, wasCalled)
+
+	assertConnWritable(t, conn, mockServer)
+
+	err = conn.Close()
+	assert.NoError(t, err)
+
+	// assert the actual connection was closed
+	assert.Error(t, clientConn.Close())
+
+	resolver.AssertExpectations(t)
+	dialer.AssertExpectations(t)
+}
+
+func TestResolvedUDPConnWriteRetry(t *testing.T) {
+	hostPort := "blahblah:34322"
+
+	mockServer, clientConn, err := newUDPConn()
+	require.NoError(t, err)
+	defer mockServer.Close()
+
+	resolver := mockResolver{}
+	resolver.
+		On("ResolveUDPAddr", "udp", hostPort).
+		Return(nil, fmt.Errorf("failed to resolve")).Once().
+		On("ResolveUDPAddr", "udp", hostPort).
+		Return(&net.UDPAddr{
```

Comment:
these addresses are repeated a lot of times, distracting from the mock instructions. Maybe pull them into mockUDPAddr1/mockUDPAddr2 vars?